src/audio/dai.c

```diff
@@ -837,6 +837,16 @@ static int dai_copy(struct comp_dev *dev)
 		 dev->direction, copy_bytes,
 		 samples / buf->stream.channels);
 
+	/* Check possibility of glitch occurrence */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK &&
+	    copy_bytes + avail_bytes < dd->period_bytes)
```

Comment:
Not following this,  I assume we are checking for a DMA buffer underrun ? If so then I would expect we want to check that avail > period_bytes (otherwise the DMA period does not have enough data)

Reply:
>we are checking for a DMA buffer underrun ? 

Yes, so when avail (bytes already loaded into DMA buffer) + number of bytes to copy (copy_bytes) must be enough to feed DMA for next period (period_bytes).

Reply:
oh, my understanding of buffer.avail is the number of bytes available to read so maybe I'm getting confused if you are meaning an upstream buffer e.g.
```
c0 -> b0 -> c1 -> b1 -> DAI
```
I'm referring to b1 here, so ```avail``` would be bytes available for DAI DMA to read for playback ?

Reply:
>oh, my understanding of buffer.avail is the number of bytes available to read

True, but it we need to extend a little proposed topology:

`
c0 -> b0 -> c1 -> b1 -> DAI -> b2 -> DMA
`

DAI component have 2 buffers, `avail` refers here to data stored in buff `b2` from such an topology. It's why we care about `free` from this buffer in playback as usual, because as usual we calculate how many bytes we can add in opposite to how many bytes already is in, it's done in following code:
```
	/* calculate minimum size to copy */
	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
		src_samples = audio_stream_get_avail_samples(&buf->stream);
		sink_samples = free_bytes / get_sample_bytes(dma_fmt);
		samples = MIN(src_samples, sink_samples);
	}
        copy_bytes = samples * get_sample_bytes(dma_fmt);
```

Reply:
is b2 a real buffer in SRAM or the DMAC internal memories/FIFO ?. If it's in SRAM then we are doing an extra copy ?

Reply:
from dai_params():
`		dd->dma_buffer = buffer_alloc(buffer_size, SOF_MEM_CAPS_DMA,
					      addr_align);
`
But it's also used in DMA (data from then are readed to DMA internal FIFO).

Reply:
why can't we pull data from b1 in your example above? I had no idea we had extra buffers.

Reply:
From commit description, where this buffer has been introduced (a year ago):
```
dma: alloc separate buffers for DMAs

Currently DMAs are using buffers on pipelines connected to either
host or dai components. This causes many problems in firmware logic
e.g. we need to preload playback pipelines with data or we need
to manually override buffer sizes, if the buffer isn't complaint
with specific DMA requirements. Also it requires from topology
creators to know some hardware specific things on each platform
like that for timer based scheduling on cAVS platforms we need
to have 3 periods in buffers connected to dai components.
This patch allocates separate buffers for DMAs, which are made
complaint by using different DMA attributes first. We no longer
need to preload playback streams, because we can start DMA
right away and just let it transfer zeroes before any valid
data will come.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>
```

Reply:
wow. The last part of the commit message is scary, we shall never transfer zeroes before valid data comes. That's against all basic principles of multi-link hardware synchronization in SoundWire.

@lgirdwood I maintain this is a bug, not a feature.

Reply:
@plbossart Fwiw 3 DMA periods is due to DW DMAC HW descriptor races with cache coherency. The descriptors all need a bit clear before they can be reused.

Reply:
@ktrzcinx just to align, is this warning is for b1  or b2 ?

Reply:
b2

Reply:
ok, so b2 needs 4 warning types. 
1) Playback FW copy() - check is FW writing to DMAC period (beyond DMA HW R ptr)?
2) Playback DMA IRQ - check is DMA now about to start reading from FW period ?
3) Capture FW copy() - check is FW reading from DMAC period (beyond DMA HW W ptr) ?
4) Capture DMA IRQ - check is DMA now about to start writing to FW period ?

Which 2 are you adding here and are the other two already upstream. We do need all 4.

Reply:
I don't understand this list 😳 I see possibility of under/over-run on playback and capture (4 cases)
1. Playback underrun -> checked in this PR (previously checked only for complete underrun, without checking already loaded samples -> comparison copy_bytes with 0)
2. Playback overrun -> when each buffer between dai and host will be full, then host show "no bytes to copy" message
3. Capture underrun -> it will be handled by host ("no bytes to copy") - here I assume possibly data rate != const are valid (slightly underrun between DAI and HOST, data stored in intermediates buffers)
4. Capture overrun -> added in this PR, there must be enough free space for for data in following period (so already `free_bytes` + what we eat `copy_bytes` must be greater than `period_bytes` )

Reply:
Your  -> mine i.e. 2 -> 1, 1 -> 2, 4->3, 3 -> 4.

It's really important that the FW and DMAC dont use the same period at the same time (and I dont see where this is validated), we only seem to validate on the whole buffer here.

Reply:
>the FW and DMAC dont use the same period at the same time

I think its 'validated' by copying no more than avail/free bytes. Avail/free bytes calculation are done by R/W pointers reads (during avail bytes calculation in dw/dma.c W ptr comes from DMA HW, R ptr is software variable, for free bytes calculation it's symmetrical, see dw/dma.c@dw_dma_get_data_size() ).

Reply:
The avail/free counters have no idea of the DMAC block size (i.e the period size). DMA will copy this asynchronously and in variable sized bursts into it's FIFO (not the DAI FIFO). The only guarantee that FW has of any coherence with the DMACs is to be sure that FW is not doing IO on the period owned by the DMAC.

Reply:
free/avail computation is safe estimation because calculation uses last known position reported by the dma.